radsat-sk/operation/services/RFileTransferService.c

```diff
@@ -0,0 +1,139 @@
+/**
+ * @file RFileTransferService.c
+ * @date February 25, 2022
+ * @author Tyrel Kostyk (tck290)
+ */
+
+#include <RFileTransferService.h>
+#include <RTransceiver.h>
+#include <RMessage.h>
+#include <hal/errors.h>
+#include <string.h>
+
+
+/***************************************************************************************************
+                                   DEFINITIONS & PRIVATE GLOBALS
+***************************************************************************************************/
+
+/** The max number of frames locally stored at one time. */
+#define MAX_FRAME_COUNT 		(200)
+
+/** Error code for internal issues regarding frame cursors (wrap-around, etc.). */
+#define ERROR_CURSOR			(-1)
+
+/** Error code for failed message wrapping. */
+#define ERROR_MESSAGE_WRAPPING	(-2)
+
+/** Struct that defines a prepared frame and its total size. */
+typedef struct _frame_t {
+	uint8_t data[TRANCEIVER_TX_MAX_FRAME_SIZE];	///> The buffer holding the prepared frame
+	uint8_t size;								///> The size (in bytes) of the entire frame
+} frame_t;
+
+
+/** Local FIFO containing frames prepared for downlink. */
+static frame_t frames[MAX_FRAME_COUNT] = { 0 };
+
+/** Cursor for writing to the frame FIFO (starts ahead of the read cursor). */
+static uint8_t frameWriteCursor = 1;
+
+/** Cursor for reading from the frame FIFO. */
+static uint8_t frameReadCursor = 0;
+
+
+/***************************************************************************************************
+                                             PUBLIC API
+***************************************************************************************************/
+
+/**
+ * Increment the internal FIFO and provide the frame at that location.
+ *
+ * Invalidates (deletes) the previous frame before moving on.
+ *
+ * @param frame Pointer to a buffer that the frame will be placed into. Set by function.
+ * @return The size of the frame placed into the buffer; 0 on error.
+ */
+uint8_t fileTransferNextFrame(uint8_t* frame) {
+
+	// ensure that there is a valid frame ahead
+	if (frames[frameReadCursor + 1].size == 0)
+		return 0;
+
+	// invalidate the current (now previous frame)
+	memset(&frames[frameReadCursor], 0, sizeof(frames[frameReadCursor]));
+
+	// increment the read cursor
+	frameReadCursor++;

```

Comment:
Need to ensure that if we hit the max frame count, we need to go back around to 0. Maybe mod it by max frame count or an if statement after or before this.

Reply:
Good catch! I just wrapped it around to 0, just like I do with the write cursor